plugins/woocommerce/src/Blocks/Patterns/PatternRegistry.php

```diff
@@ -0,0 +1,197 @@
+<?php
+namespace Automattic\WooCommerce\Blocks\Patterns;
+
+use Automattic\WooCommerce\Admin\Features\Features;
+
+/**
+ * PatternRegistry class.
+ */
+class PatternRegistry {
+	const SLUG_REGEX            = '/^[A-z0-9\/_-]+$/';
+	const COMMA_SEPARATED_REGEX = '/[\s,]+/';
+
+	/**
+	 * Register a block pattern.
+	 *
+	 * @param string $source The pattern source.
+	 * @param array  $pattern_data The pattern data.
+	 * @param array  $dictionary The patterns' dictionary.
+	 *
+	 * @return void
+	 */
+	public function register_block_pattern( $source, $pattern_data, $dictionary ) {
+		if ( empty( $pattern_data['slug'] ) ) {
+			_doing_it_wrong(
+				'register_block_patterns',
+				esc_html(
+					sprintf(
+					/* translators: %s: file name. */
+						__( 'Could not register pattern "%s" as a block pattern ("Slug" field missing)', 'woocommerce' ),
+						$source
+					)
+				),
+				'6.0.0'
+			);
+			return;
+		}
+
+		if ( ! preg_match( self::SLUG_REGEX, $pattern_data['slug'] ) ) {
+			_doing_it_wrong(
+				'register_block_patterns',
+				esc_html(
+					sprintf(
+					/* translators: %1s: file name; %2s: slug value found. */
+						__( 'Could not register pattern "%1$s" as a block pattern (invalid slug "%2$s")', 'woocommerce' ),
+						$source,
+						$pattern_data['slug']
+					)
+				),
+				'6.0.0'
+			);
+			return;
+		}
+
+		if ( \WP_Block_Patterns_Registry::get_instance()->is_registered( $pattern_data['slug'] ) ) {
+			return;
+		}
+
+		if ( isset( $pattern_data['featureFlag'] ) && '' !== $pattern_data['featureFlag'] && ! Features::is_enabled( $pattern_data['featureFlag'] ) ) {
+			return;
+		}
```

Comment:
Just out of curiosity, can you give more details on the first two checks? 
Could a pattern have a feature flag value?

Reply:
Sure, it was added here https://github.com/woocommerce/woocommerce/pull/46856/

Reply:
TIL! Thanks!